src/ncp/posix/infra_if.cpp

```diff
@@ -0,0 +1,158 @@
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#define OTBR_LOG_TAG "INFRAIF"
+
+#include "infra_if.hpp"
+
+#include <ifaddrs.h>
+#ifdef __linux__
+#include <linux/netlink.h>
+#include <linux/rtnetlink.h>
+#endif
+#include <netinet/icmp6.h>
+#include <sys/ioctl.h>
+
+#include "utils/socket_utils.hpp"
+
+namespace otbr {
+
+otbrError InfraIf::Dependencies::SetInfraIf(uint32_t                       aInfraIfIndex,
+                                            bool                           aIsRunning,
+                                            const std::vector<Ip6Address> &aIp6Addresses)
+{
+    OTBR_UNUSED_VARIABLE(aInfraIfIndex);
+    OTBR_UNUSED_VARIABLE(aIsRunning);
+    OTBR_UNUSED_VARIABLE(aIp6Addresses);
+    return OTBR_ERROR_NONE;
+}
+
+InfraIf::InfraIf(Dependencies &aDependencies)
+    : mDeps(aDependencies)
+    , mInfraIfIndex(0)
+{
+}
+
+void InfraIf::Init(void)
+{
+}
+
+void InfraIf::Deinit(void)
+{
+    mInfraIfIndex = 0;
+}
+
+void InfraIf::SetInfraIf(const char *aIfName)
+{
+    std::vector<Ip6Address> addresses;
+
+    VerifyOrDie(aIfName != nullptr && strlen(aIfName) > 0, "Invalid infra interface name");
+    VerifyOrDie(strnlen(aIfName, sizeof(mInfraIfName)) < sizeof(mInfraIfName), "infra interface name is too long!");
+    strcpy(mInfraIfName, aIfName);
+
+    mInfraIfIndex = if_nametoindex(aIfName);
+    VerifyOrDie(mInfraIfIndex != 0, "Failed to get the index for infra interface");
+
+    GetAddresses(addresses);
+
+    SuccessOrDie(mDeps.SetInfraIf(mInfraIfIndex, IsRunning(addresses), addresses), "Failed to set infra-if!");
+}
+
+bool InfraIf::IsRunning(const std::vector<Ip6Address> &aAddrs) const
+{
+    return mInfraIfIndex ? ((GetFlags() & IFF_RUNNING) && HasLinkLocalAddress(aAddrs)) : false;
+}
+
+uint32_t InfraIf::GetFlags(void) const
+{
+    int          sock;
+    struct ifreq ifReq;
+    uint32_t     flags = 0;
+
+    sock = SocketWithCloseExec(AF_INET6, SOCK_DGRAM, IPPROTO_IP, kSocketBlock);
+    VerifyOrDie(sock != -1, otbrErrorString(OTBR_ERROR_ERRNO));
+
+    memset(&ifReq, 0, sizeof(ifReq));
+    static_assert(sizeof(ifReq.ifr_name) >= sizeof(mInfraIfName), "mInfraIfName is not of appropriate size.");
+    strcpy(ifReq.ifr_name, mInfraIfName);
+
+    if (ioctl(sock, SIOCGIFFLAGS, &ifReq) == -1)
+    {
+        otbrLogCrit("The infra link %s may be lost. Exiting.", mInfraIfName);
+        DieNow(otbrErrorString(OTBR_ERROR_ERRNO));
+    }
+    flags = static_cast<uint32_t>(ifReq.ifr_flags);
+
+    close(sock);
+    return flags;
+}
+
+void InfraIf::GetAddresses(std::vector<Ip6Address> &aAddrs)
+{
+    struct ifaddrs *ifAddrs = nullptr;
+
+    if (getifaddrs(&ifAddrs) < 0)
+    {
+        otbrLogCrit("failed to get netif addresses: %s", strerror(errno));
+        DieNow(otbrErrorString(OTBR_ERROR_ERRNO));
```

Comment:
Ditto. I'm kind of concerned about using `Die` in such places, although it's already there in OT's infra_if.cpp